File: app/src/main/java/com/bitflaker/lucidsourcekit/charts/RangeProgress.java
package com.bitflaker.lucidsourcekit.charts;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.graphics.Shader;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;

import com.bitflaker.lucidsourcekit.R;
import com.bitflaker.lucidsourcekit.general.Tools;

public class RangeProgress extends View {
    private final Paint dataLabelPaint = new Paint(Paint.ANTI_ALIAS_FLAG | Paint.DITHER_FLAG);
    private final Paint dataLinePaint = new Paint();
    private final Paint dataLinePaintBackground = new Paint();
    private float value;
    private float xMax;
    private float textHeight;
    private int iconSize;
    private Bitmap icon;
    private Drawable iconDrawable;
    private String text;
    private String label;
    private int minHeight = 0;
    private final Rect textBounds = new Rect();
    private int[] progressColors;
    private int[] textColors;
    private float[] positions;
    private float percentage;
    private float lastWidth = 0;
    private LinearGradient lineShader;
    private LinearGradient textShader;
    private boolean initialized;

    public RangeProgress(Context context){
        super(context);
        setup();
    }

    public RangeProgress(Context context, AttributeSet attributeSet){
        super(context, attributeSet);
        setup();
    }

    private void setup() {
        dataLinePaintBackground.setColor(Tools.getAttrColor(R.attr.backgroundColor, getContext().getTheme()));
        dataLinePaintBackground.setStrokeCap(Paint.Cap.ROUND);
        dataLinePaintBackground.setAntiAlias(true);
        dataLinePaintBackground.setStrokeWidth(0);
        dataLinePaint.setColor(Tools.getAttrColor(R.attr.colorSecondary, getContext().getTheme()));
        dataLinePaint.setStrokeCap(Paint.Cap.ROUND);
        dataLinePaint.setAntiAlias(true);
        dataLinePaint.setStrokeWidth(0);
        dataLabelPaint.setColor(Tools.getAttrColor(R.attr.primaryTextColor, getContext().getTheme()));
        dataLabelPaint.setTextSize((int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 14, getResources().getDisplayMetrics()));
        dataLabelPaint.setTextAlign(Paint.Align.LEFT);
        dataLabelPaint.setFakeBoldText(true);
        dataLabelPaint.setAntiAlias(true);
        progressColors = new int[] {Tools.getAttrColor(R.attr.colorSecondary, getContext().getTheme()), Tools.getAttrColor(R.attr.backgroundColor, getContext().getTheme())};
        textColors = new int[] { Tools.getAttrColor(R.attr.colorOnSecondary, getContext().getTheme()), Tools.manipulateAlpha(Tools.getAttrColor(R.attr.secondaryTextColor, getContext().getTheme()), 0.7f) };
        positions = new float[] { 0f, 0f };
    }

    public void setBackgroundAttrColor(int color){
        dataLinePaintBackground.setColor(Tools.getAttrColor(color, getContext().getTheme()));
        progressColors = new int[] {Tools.getAttrColor(R.attr.colorSecondary, getContext().getTheme()), Tools.getAttrColor(color, getContext().getTheme())};
        invalidate();
    }

    public void setData(float maxValue, float value, String label, Drawable icon, String text) {
        this.value = value;
        this.label = label;
        this.text = text;
        if(this.text != null && this.text.equals("NaN")){
            this.text = "-";
        }
        this.xMax = maxValue;
        this.iconDrawable = icon;
        this.textHeight = 0;
        percentage = value / xMax;
        if(maxValue == 0 || Float.isNaN(value)){
            percentage = 0;
            this.text = "-";
        }
        initialized = true;
        invalidate();
    }

    public boolean isInitialized() {
        return initialized;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        float minRadMargin = getHeight() / 2.0f;
        float textMargin = 27;

        if(lineShader == null || textShader == null || positions[0] != percentage || positions[1] != percentage || lastWidth != getWidth()){
            positions[0] = percentage;
            positions[1] = percentage;
            lineShader = new LinearGradient(0f, 0f, getWidth(), 0f, progressColors, positions, Shader.TileMode.MIRROR);
            textShader = new LinearGradient(0f, 0f, getWidth(), 0f, textColors, positions, Shader.TileMode.MIRROR);
        }

        if(dataLinePaintBackground.getStrokeWidth() == 0){
            dataLinePaintBackground.setStrokeWidth(Tools.dpToPx(getContext(), Tools.pxToDp(getContext(), getHeight())));
        }

        if(dataLinePaint.getStrokeWidth() == 0){
            dataLinePaint.setStrokeWidth(Tools.dpToPx(getContext(), Tools.pxToDp(getContext(), getHeight())));
        }

        if(icon == null && iconDrawable != null){
            iconSize = (int) (getHeight() / 5.0f * 3.25);
            icon = drawableToBitmap(iconDrawable, iconSize);
        }

        dataLinePaintBackground.setShader(lineShader);
        canvas.drawLine(minRadMargin, getHeight() / 2.0f, getWidth()-minRadMargin, getHeight() / 2.0f, dataLinePaintBackground);

        if(label != null && label.length() > 0) {
            dataLabelPaint.getTextBounds(label, 0, label.length(), textBounds);
            dataLabelPaint.setShader(textShader);
            canvas.drawText(label, textMargin, getHeight() / 2.0f - textBounds.exactCenterY(), dataLabelPaint);
        }

        if(icon != null) {
            float begin = getWidth() - iconSize - (getHeight() - iconSize) / 2.0f - 5; // margin of 5 to the right in order to see the end a bit better if the bar is not totally filled
            float end = begin + iconSize;
            float current = percentage * getWidth();
            float iconFillPercentage = 1;
            if(current > begin && current < end) {
                float within = current - begin;
                iconFillPercentage = within / iconSize;
            }
            else if(current <= begin) {
                iconFillPercentage = 0;
            }
            canvas.drawBitmap(addGradient(icon, iconSize, iconFillPercentage), begin, getHeight()/2.0f-iconSize/2.0f, dataLabelPaint);
        }
        else if(text != null && text.length() > 0) {
            dataLabelPaint.getTextBounds(text, 0, text.length(), textBounds);
            canvas.drawText(text, getWidth() - textMargin - textBounds.width(), getHeight()/2.0f - textBounds.exactCenterY(), dataLabelPaint);
        }

        lastWidth = getWidth();
    }

    public static Bitmap drawableToBitmap (Drawable drawable, int size) {
        if (drawable instanceof BitmapDrawable) {
            return ((BitmapDrawable)drawable).getBitmap();
        }

        Bitmap bitmap = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, size, size);
        drawable.draw(canvas);

        return bitmap;
    }

    public Bitmap addGradient(Bitmap src, int size, float pos) {
        Bitmap result = Bitmap.createBitmap(size,size, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(result);

        canvas.drawBitmap(src, 0, 0, null);

        Paint paint = new Paint();
        LinearGradient shader = new LinearGradient(0,0,size,0, textColors, new float[] { pos, pos }, Shader.TileMode.CLAMP);
        paint.setShader(shader);
        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
        canvas.drawRect(0,0,size,size,paint);

        return result;
    }
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/views/RangeProgress.kt
package com.bitflaker.lucidsourcekit.views

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.Rect
import android.graphics.Shader
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View
import com.bitflaker.lucidsourcekit.R
import com.bitflaker.lucidsourcekit.utils.Tools.manipulateAlpha
import com.bitflaker.lucidsourcekit.utils.attrColor
import com.bitflaker.lucidsourcekit.utils.spToPx
import com.bitflaker.lucidsourcekit.utils.toBitmap
import kotlin.Int
import kotlin.String
import kotlin.floatArrayOf
import kotlin.intArrayOf
import androidx.core.graphics.createBitmap
import com.bitflaker.lucidsourcekit.utils.getTextBounds
import kotlin.math.max
import kotlin.math.min

class RangeProgress @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {
    // Pre-configured paints
    private val dataLabelPaint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.DITHER_FLAG).apply {
        color = context.attrColor(R.attr.primaryTextColor)
        isFakeBoldText = true
        textAlign = Paint.Align.LEFT
        textSize = 14f.spToPx
    }
    private val dataLinePaint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.DITHER_FLAG).apply {
        color = context.attrColor(R.attr.colorSurfaceContainer)
        strokeCap = Paint.Cap.ROUND
        strokeWidth = 0f
    }
    private val iconGradientPaint = Paint().apply {
        xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
    }

    // Configurable view specific fields
    var label: String? = null
        set(value) {
            field = value
            textCenterY = value?.let {
                val labelBounds = Rect()
                dataLabelPaint.getTextBounds(value, labelBounds)
                labelBounds.exactCenterY()
            } ?: 0f
        }

    var icon: Drawable? = null
        set(value) {
            field = value
            if (value == null)  {
                iconBitmap = null
            }
            else {
                text = null
            }
        }

    var text: String? = null
        set(value) {
            field = value
            if (value != null) {
                icon = null
                if (value == "NaN") {
                    field = "-"
                }
            }
            dataLabelPaint.getTextBounds(text ?: "", textBounds)
        }

    var labelInsetX = 27f
    var iconInsetX = 5f

    // Internal values
    private var progressColors = intArrayOf(
        context.attrColor(R.attr.colorTertiary),
        context.attrColor(R.attr.colorSurfaceContainer)
    )
    private var textColors = intArrayOf(
        context.attrColor(R.attr.colorOnTertiary),
        manipulateAlpha(context.attrColor(R.attr.secondaryTextColor), 0.7f)
    )
    private var value = 0f
    private var xMax = 0f
    private var positions = floatArrayOf(0f, 0f)
    private var percentage = 0f
    private var iconBitmap: Bitmap? = null
    private var iconGradientBitmap = createBitmap(1, 1)
    private var iconGradientCanvas = Canvas(iconGradientBitmap)
    private val iconGradientMatrix = Matrix()
    private val textBounds = Rect()
    private var textCenterY = 0f
    private var iconSize: Int = 0
        set(value) {
            field = value
            iconBitmap = icon?.toBitmap(value)
            iconGradientBitmap = createBitmap(value, value)
            iconGradientCanvas = Canvas(iconGradientBitmap)
            iconGradientPaint.shader = LinearGradient(
                0f,
                0f,
                value.toFloat(),
                0f,
                textColors,
                floatArrayOf(0f, 0f),
                Shader.TileMode.CLAMP
            )
        }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val inset = height / 2.0f

        // Draw the progress and track line
        canvas.drawLine(
            inset,
            height / 2.0f,
            width - inset,
            height / 2.0f,
            dataLinePaint
        )

        // Draw the label if present
        label?.let {
            canvas.drawText(
                it,
                labelInsetX,
                height / 2.0f - textCenterY,
                dataLabelPaint
            )
        }

        // Draw the icon on the right end of the view if present
        iconBitmap?.let {
            // Get the x position of the start of the icon
            val xStart = width - iconSize - (height - iconSize) / 2.0f - iconInsetX

            // Get the percentage of the icon being on top of the filled progress and update the icon gradient
            val progressStopX = percentage * width
            val iconFillPercentage = min(iconSize.toFloat(), max(0f, progressStopX - xStart)) / iconSize
            updateIconGradient(it, iconFillPercentage)

            // Draw the icon vertically centered to the target x-position
            canvas.drawBitmap(
                iconGradientBitmap,
                xStart,
                (height - iconSize) / 2.0f,
                dataLabelPaint
            )
        }

        // Draw the text on the right end of the view if present
        text?.let {
            canvas.drawText(
                it,
                width - labelInsetX - textBounds.width(),
                height / 2.0f - textBounds.exactCenterY(),
                dataLabelPaint
            )
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val width = MeasureSpec.getSize(widthMeasureSpec).toFloat()
        val height = MeasureSpec.getSize(heightMeasureSpec).toFloat()

        // Update paint strokes to new height
        dataLinePaint.strokeWidth = height

        // Update shaders to new width
        dataLinePaint.shader = getLineShader(width)
        dataLabelPaint.shader = getTextShader(width)

        // Update icon size to new view height
        iconSize = (height / 5.0f * 3.25).toInt()

        invalidate()
    }

    private fun getTextShader(width: Float): LinearGradient = LinearGradient(
        0f,
        0f,
        width,
        0f,
        textColors,
        positions,
        Shader.TileMode.MIRROR
    )

    private fun getLineShader(width: Float): LinearGradient = LinearGradient(
        0f,
        0f,
        width,
        0f,
        progressColors,
        positions,
        Shader.TileMode.MIRROR
    )

    private fun updateIconGradient(src: Bitmap, percentage: Float) {
        // Translate the matrix of the gradient to the percentage of the icon width on top of the progress
        iconGradientMatrix.reset()
        iconGradientMatrix.setTranslate(iconSize * percentage, 0f)
        (iconGradientPaint.shader as LinearGradient).setLocalMatrix(iconGradientMatrix)

        // Re-draw the icon to the bitmap with the updated gradient
        iconGradientCanvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR)
        iconGradientCanvas.drawBitmap(src, 0f, 0f, null)
        iconGradientCanvas.drawRect(0f, 0f, iconSize.toFloat(), iconSize.toFloat(), iconGradientPaint)
    }

    fun setBackgroundAttrColor(colorRes: Int) {
        dataLinePaint.color = context.attrColor(colorRes)
        progressColors = intArrayOf(
            context.attrColor(R.attr.colorTertiary),
            dataLinePaint.color
        )

        invalidate()
    }

    fun setValue(value: Float, max: Float) {
        this.value = value
        this.xMax = max

        percentage = value / xMax
        if (xMax == 0f || value.isNaN()) {
            percentage = 0f
            text = "-"
        }

        positions[0] = percentage
        positions[1] = percentage
        dataLinePaint.shader = getLineShader(width.toFloat())
        dataLabelPaint.shader = getTextShader(width.toFloat())

        invalidate()
    }
}